### PR TITLE
Content shape alternate targeting a specific item with ContentItemId (#15572)

### LINKED ISSUE #15572: Content shape alternate targeting a specific item with ContentItemId
### Is your feature request related to a problem? Please describe.
You can override the `Content` shape of a given content item with the `Content__[Id]` alternate, which is added in [`OrchardCore.Contents.Shapes`](https://github.com/OrchardCMS/OrchardCore/blob/d524386b2f792f35773324ae482247e80a944266/src/OrchardCore.Modules/OrchardCore.Contents/Shapes.cs#L37). This is useful, but since it uses `ContentItem.Id`, i.e. the numeric ID of the item in the DB, it's not portable: an override will only apply with one specific DB, and if you move the content item (like import it) to another DB, it won't.

### Describe the solution you'd like
Add a new alternate, `Content__[ContentItemId]`, that uses the alphanumeric and portable `ContentItem.ContentItemId`. Also document it under https://docs.orchardcore.net/en/latest/docs/reference/modules/Templates/#content-templates.

We can keep the current alternate for backward compatibility, but it shouldn't really be used, so let's keep it undocumented.

### Describe alternatives you've considered
Changing the existing alternate, but that would be a breaking change.

### PR DESCRIPTION
**Pull request by dmourtzoukos**
This pull request enhances the templating system for content items by adding support for templates based on `ContentItemId` and updating the documentation to reflect these changes. The most important changes include the addition of new template alternates in the `Shapes.cs` file and the corresponding updates in the documentation.

### Enhancements to templating system:

* [`src/OrchardCore.Modules/OrchardCore.Contents/Shapes.cs`](diffhunk://#diff-3e09c2fd05fbf4a41031b6243e3a14ddcc9a0160056fa512b9c1a4cfd74be33aR35-R46): Added new template alternates for `ContentItemId` and `DisplayType` combined with `ContentItemId`. This allows for more granular control over content item templates.

### Documentation updates:

* [`src/docs/reference/modules/Templates/README.md`](diffhunk://#diff-3c3d5402b80eb850641afd07e09d5aca737dfe03de6889966606e8f7710e8135R54-R99): Updated the documentation to include examples and descriptions for the new template alternates based on `ContentItemId` and `DisplayType` combined with `ContentItemId`.

Fixes #15572.

### CODE REVIEW

**Comment by dmourtzoukos** [comment]
Fixes following issue:
https://github.com/OrchardCMS/OrchardCore/issues/15572

File: src/docs/reference/modules/Templates/README.md [hunk 1]
```diff
@@ -51,6 +51,52 @@ For instance, when a content item is displayed in a list, the `Summary` display
...
+| Template | Filename|
+| --------- | ------------ |
+| `Content__12` | `Content-12.cshtml` 
+
+### `Content_[DisplayType]__[Id]`
+
+This template is called when displaying a content item with a specific display type and Id.
+For instance, when a content item is displayed in a list, the `Summary` display type is commonly used.
+
+Note: This template was supported on Orcard Core version 1.X and will probably be removed from future versions. It is strongly recommended to use the corresponding template that utilises the ContentItemId.
```
Comment (hishamco) [inline src/docs/reference/modules/Templates/README.md:92]: Duplicate

File: src/docs/reference/modules/Templates/README.md [hunk 2]
```diff
@@ -51,6 +51,52 @@ For instance, when a content item is displayed in a list, the `Summary` display
...
+
+| Template | Filename|
+| --------- | ------------ |
+| `Content_Summary__42n8364bk2azcy1yd71gnrntwf` | `Content-42n8364bk2azcy1yd71gnrntwf.Summary.cshtml` 
+
+### `Content__[Id]`
+
+This template is called when displaying a content item with a specific Id.
+
+Note: This template was supported on Orcard Core version 1.X and will probably be removed from future versions. It is strongly recommended to use the corresponding template that utilises the ContentItemId.
```
Comment (hishamco) [inline src/docs/reference/modules/Templates/README.md:79]:
```suggestion
Note: This template was supported on Orchard Core version 1.X and will probably be removed from future versions. It is strongly recommended to use the corresponding template that utilizes the ContentItemId.
```

Comment (hishamco) [inline src/docs/reference/modules/Templates/README.md:79]: Orchard Core or Orchard 1?

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by Piedone** [comment]
I'm out until Harvest and won't be able to review, trust your judgement @hishamco.

File: src/docs/reference/modules/Templates/README.md [hunk 3]
```diff
@@ -76,7 +76,7 @@ For instance, when a content item is displayed in a list, the `Summary` display
 
 This template is called when displaying a content item with a specific Id.
 
-Note: This template was supported on Orcard Core version 1.X and will probably be removed from future versions. It is strongly recommended to use the corresponding template that utilises the ContentItemId.
+Note: This template was supported on Orchard version 1.X and will probably be removed from future versions. It is strongly recommended to use the corresponding template that utilizes the ContentItemId.
```
Comment (hishamco) [inline src/docs/reference/modules/Templates/README.md:79]: > and will probably be removed from future versions

Your source?

Comment (hishamco) [reply src/docs/reference/modules/Templates/README.md:79]: @sebastienros do you remember if this is still supported in Orchard CMS?

/cc @BenedekFarkas

Comment (sebastienros) [reply src/docs/reference/modules/Templates/README.md:79]: @hishamco I think it's just quoting the issue, which states that there is no reason to use Id as it's not stable (I agree). So we could remove that later but not right now as we need to provide a transition window (let users migrate to Id). So maybe file an issue to remove this and assign to the next release? (not a .x one).

Comment (sebastienros) [reply src/docs/reference/modules/Templates/README.md:79]: @dmourtzoukos Don't we have alias and route already generating these alternates? That could be a nicer option in most cases. Not sure how many items you would have that are not accessible in either way. But I have nothing against this PR, if not that it's yet another shape that allocates quite a bit, but we'll optimize this eventually.

Comment (MichaelPetrinolis) [reply src/docs/reference/modules/Templates/README.md:79]: @sebastienros there is no documentation about alternates for Alias or Autoroute (and could not find any shape table provider in code). My concern here has to do with localized items, with any of the three approaches -Alias, Route and ContentItemId-.
I mean, how could somebody override the template for a localization set? Defining templates per ContentItemId and using a shared liquid? Maybe is better to also introduce alternates per LocalizationSet?

Comment (barthamark) [reply src/docs/reference/modules/Templates/README.md:79]: As part of the Orchard Harvest Hackathon, I'm adding alternates for the alias and the routes. The issue specific to this is #2866 and the pull request is #16722

File: src/docs/reference/modules/Templates/README.md
Comment (Piedone) [inline src/docs/reference/modules/Templates/README.md:1]: Please add a note about this to https://github.com/OrchardCMS/OrchardCore/blob/main/src/docs/releases/2.1.0.md. See https://docs.orchardcore.net/en/latest/guides/contributing/contributing-documentation/.

**Review by Piedone — CHANGES_REQUESTED** [review]
I didn't review the PR, but am only adding the release notes note.

**Comment by hishamco** [comment]
Please resolve the conflict

**Comment by dmourtzoukos** [comment]
> Please resolve the conflict

Conflicts have been resolved.

**Comment by Piedone** [comment]
I'm not reviewing but @hishamco please finish yours.

**Comment by MikeAlhayek** [comment]
@dmourtzoukos If you please add notes about this change in https://github.com/OrchardCMS/OrchardCore/blob/main/src/docs/releases/3.0.0.md I will merge it for you. Please add the note and request my review so I can merge it.

**Comment by sebastienros** [comment]
Do we really need this since we have alias and route shapes? I am thinking this would just create extraneous alternates for every single content item that would probably be unnecessary.

**Comment by MikeAlhayek** [comment]
@sebastienros same feeling here. However, others seems to want it. If it is not really needed, I rather not add it

**Comment by sebastienros** [comment]
> others seems to want it

The issues were created before we had alias/route alternates, maybe we should ask again.

**Comment by MichaelPetrinolis** [comment]
You are right, now that we have alternates by alias this PR has no added value.
Do you agree to modify the alternate based on the localization set, so that a simple template can be applied to a specific content item on any locale? This cannot be achieved by alias or autoroute alternates, as they are unique

**Comment by sebastienros** [comment]
@MichaelPetrinolis can you file an issue? Also here is a previous mention of this, check the whole thread there are other alternatives. https://github.com/OrchardCMS/OrchardCore/issues/5033#issuecomment-565657974

It's not that easy though, perf implications (even more that this PR)

**Comment by Piedone** [comment]
Yeah, this is not really needed anymore, but thank you @dmourtzoukos! Sorry that things shifted around in the meantime, but we'd welcome any further improvements/issue fixes from you in the future.

Closed https://github.com/OrchardCMS/OrchardCore/issues/15572.

**Pull request closed without merging**